File: source/hw_config.c
#include "hw_config.h"
#include <stdio.h>
#include <math.h>

void Init_CIM_Macro(CIM_Macro *macro, int AL, int PC, int SCR, int ICW, int WUW, int WEIGHT_COL, int INPUT_WIDTH, int RESULT_WIDTH, int DIRECTION, int DATA_TYPE, int WEIGHT_WIDTH, int WEIGHT_ROW) {
    macro->AL = AL;
    macro->PC = PC;
    macro->SCR = SCR;
    macro->ICW = ICW;
    macro->WUW = WUW;
    macro->DIRECTION = DIRECTION;
    macro->DATA_TYPE = DATA_TYPE;
    macro->WEIGHT_WIDTH = WEIGHT_WIDTH;
    macro->WEIGHT_ROW = WEIGHT_ROW;
    macro->WEIGHT_COL = WEIGHT_COL;
    macro->INPUT_WIDTH = INPUT_WIDTH;
    macro->RESULT_WIDTH = RESULT_WIDTH;
}

void Init_Micro_Arch(Micro_Arch *arch, CIM_Macro *macro, int MACRO_ROW, int MACRO_COL, int IS_DEPTH, int OS_DEPTH, int FREQ, int BUS_WIDTH, int PIPELINE_STAGES) {
    arch->MACRO_ROW = MACRO_ROW;
    arch->MACRO_COL = MACRO_COL;
    arch->AL = MACRO_COL * macro->AL;
    arch->PC = MACRO_ROW * macro->PC;
    arch->SCR = macro->SCR;
    arch->ICW = MACRO_COL * macro->ICW;
    arch->WUW = MACRO_COL * macro->WUW;

    arch->DIRECTION = macro->DIRECTION;
    arch->DATA_TYPE = macro->DATA_TYPE;
    arch->WEIGHT_WIDTH = macro->WEIGHT_WIDTH;
    arch->WEIGHT_ROW = macro->WEIGHT_ROW;
    arch->WEIGHT_COL = macro->WEIGHT_COL;
    arch->INPUT_WIDTH = macro->INPUT_WIDTH;
    arch->RESULT_WIDTH = macro->RESULT_WIDTH;
    if (macro->DIRECTION == 0){
        // arch->CIM_WIDTH = macro->WEIGHT_COL * arch->AL;
        arch->CIM_DEPTH = macro->WEIGHT_ROW * macro->SCR * arch->PC;
    }
    else {
        // arch->CIM_DEPTH = macro->WEIGHT_COL * arch->AL * macro->SCR;
        arch->CIM_WIDTH = macro->WEIGHT_ROW * arch->PC;
    }
    arch->CIM_WIDTH = arch->WUW;
    arch->CIM_SIZE = arch->CIM_WIDTH * arch->CIM_DEPTH;

    // 初始化 Cache
    arch->IS_WIDTH = arch->ICW;
    arch->IS_DEPTH = IS_DEPTH;
    arch->OS_WIDTH = (macro->RESULT_WIDTH+(int)(log2(arch->MACRO_COL)))*arch->PC;
    arch->OS_DEPTH = OS_DEPTH;
    arch->IS_SIZE = arch->IS_WIDTH * arch->IS_DEPTH;
    arch->OS_SIZE = arch->OS_WIDTH * arch->OS_DEPTH;

    // 初始化其他参数
    arch->FREQ = FREQ;
    arch->BUS_WIDTH = BUS_WIDTH;
    arch->PIPELINE_STAGES = PIPELINE_STAGES;
}

void Print_CIM_Macro(const CIM_Macro *macro) {
    printf("CIM_Macro:\n");
    printf("AL: %d\n", macro->AL);
    printf("PC: %d\n", macro->PC);
    printf("SCR: %d\n", macro->SCR);
    printf("ICW: %d\n", macro->ICW);
    printf("WUW: %d\n", macro->WUW);
    printf("WEIGHT_COL: %d\n", macro->WEIGHT_COL);
    printf("INPUT_WIDTH: %d\n", macro->INPUT_WIDTH);
    printf("RESULT_WIDTH: %d\n", macro->RESULT_WIDTH);

    printf("DIRECTION: %d\n", macro->DIRECTION);
    printf("DATA_TYPE: %d\n", macro->DATA_TYPE);
    printf("WEIGHT_WIDTH: %d\n", macro->WEIGHT_WIDTH);
    printf("WEIGHT_ROW: %d\n", macro->WEIGHT_ROW);
    printf("\n");
}

void Print_Micro_Arch(const Micro_Arch *arch) {
    printf("Micro_Arch:\n");
    printf("MACRO_ROW: %d\n", arch->MACRO_ROW);
    printf("MACRO_COL: %d\n", arch->MACRO_COL);
    printf("AL: %d\n", arch->AL);
    printf("PC: %d\n", arch->PC);
    printf("SCR: %d\n", arch->SCR);
    printf("ICW: %d\n", arch->ICW);
    printf("WUW: %d\n", arch->WUW);
    printf("DIRECTION: %d\n", arch->DIRECTION);
    printf("DATA_TYPE: %d\n", arch->DATA_TYPE);
    printf("WEIGHT_WIDTH: %d\n", arch->WEIGHT_WIDTH);
    printf("WEIGHT_ROW: %d\n", arch->WEIGHT_ROW);
    printf("WEIGHT_COL: %d\n", arch->WEIGHT_COL);
    printf("INPUT_WIDTH: %d\n", arch->INPUT_WIDTH);
    printf("RESULT_WIDTH: %d\n", arch->RESULT_WIDTH);
    printf("CIM_WIDTH: %d\n", arch->CIM_WIDTH);
    printf("CIM_DEPTH: %d\n", arch->CIM_DEPTH);
    printf("CIM_SIZE: %d\n", arch->CIM_SIZE);
    printf("IS_WIDTH: %d\n", arch->IS_WIDTH);
    printf("IS_DEPTH: %d\n", arch->IS_DEPTH);
    printf("OS_WIDTH: %d\n", arch->OS_WIDTH);
    printf("OS_DEPTH: %d\n", arch->OS_DEPTH);
    printf("IS_SIZE: %d\n", arch->IS_SIZE);
    printf("OS_SIZE: %d\n", arch->OS_SIZE);
    printf("FREQ: %d\n", arch->FREQ);
    printf("BUS_WIDTH: %d\n", arch->BUS_WIDTH);
    printf("PIPELINE_STAGES: %d\n", arch->PIPELINE_STAGES);
    printf("\n");
}

File: evaluator/hw_config.py
import math

class CIMMacro:
    def __init__(self, AL, PC, SCR, ICW, WUW, WEIGHT_COL, INPUT_WIDTH, RESULT_WIDTH, DIRECTION, DATA_TYPE, WEIGHT_WIDTH, WEIGHT_ROW):
        self.AL = AL
        self.PC = PC
        self.SCR = SCR
        self.ICW = ICW
        self.WUW = WUW
        self.WEIGHT_COL = WEIGHT_COL
        self.INPUT_WIDTH = INPUT_WIDTH
        self.RESULT_WIDTH = RESULT_WIDTH
        self.DIRECTION = DIRECTION
        self.DATA_TYPE = DATA_TYPE
        self.WEIGHT_WIDTH = WEIGHT_WIDTH
        self.WEIGHT_ROW = WEIGHT_ROW

class MicroArch:
    def __init__(self, macro, MACRO_ROW, MACRO_COL, IS_DEPTH, OS_DEPTH, FREQ, BUS_WIDTH, PIPELINE_STAGES):
        self.MACRO_ROW = MACRO_ROW
        self.MACRO_COL = MACRO_COL
        self.AL = MACRO_COL * macro.AL
        self.PC = MACRO_ROW * macro.PC
        self.SCR = macro.SCR
        self.ICW = MACRO_COL * macro.ICW
        self.WUW = MACRO_COL * macro.WUW
        self.DIRECTION = macro.DIRECTION
        self.DATA_TYPE = macro.DATA_TYPE
        self.WEIGHT_WIDTH = macro.WEIGHT_WIDTH
        self.WEIGHT_ROW = macro.WEIGHT_ROW
        self.WEIGHT_COL = macro.WEIGHT_COL
        self.INPUT_WIDTH = macro.INPUT_WIDTH
        self.RESULT_WIDTH = macro.RESULT_WIDTH
        
        if macro.DIRECTION == 0:
            self.CIM_DEPTH = macro.WEIGHT_ROW * macro.SCR * self.PC
        else:
            self.CIM_WIDTH = macro.WEIGHT_ROW * self.PC
        
        self.CIM_WIDTH = self.WUW
        self.CIM_SIZE = self.CIM_WIDTH * self.CIM_DEPTH

        self.IS_WIDTH = self.ICW
        self.IS_DEPTH = IS_DEPTH
        self.OS_WIDTH = (macro.RESULT_WIDTH + int(math.log2(self.MACRO_COL))) * self.PC
        self.OS_DEPTH = OS_DEPTH
        self.IS_SIZE = self.IS_WIDTH * self.IS_DEPTH
        self.OS_SIZE = self.OS_WIDTH * self.OS_DEPTH

        self.FREQ = FREQ
        self.BUS_WIDTH = BUS_WIDTH
        self.PIPELINE_STAGES = PIPELINE_STAGES

    def print_macro(self, macro):
        print("CIMMacro:")
        print("AL:", macro.AL)
        print("PC:", macro.PC)
        print("SCR:", macro.SCR)
        print("ICW:", macro.ICW)
        print("WUW:", macro.WUW)
        print("WEIGHT_COL:", macro.WEIGHT_COL)
        print("INPUT_WIDTH:", macro.INPUT_WIDTH)
        print("RESULT_WIDTH:", macro.RESULT_WIDTH)
        print("DIRECTION:", macro.DIRECTION)
        print("DATA_TYPE:", macro.DATA_TYPE)
        print("WEIGHT_WIDTH:", macro.WEIGHT_WIDTH)
        print("WEIGHT_ROW:", macro.WEIGHT_ROW)
        print()

    def print_arch(self):
        print("MicroArch:")
        print("MACRO_ROW:", self.MACRO_ROW)
        print("MACRO_COL:", self.MACRO_COL)
        print("AL:", self.AL)
        print("PC:", self.PC)
        print("SCR:", self.SCR)
        print("ICW:", self.ICW)
        print("WUW:", self.WUW)
        print("DIRECTION:", self.DIRECTION)
        print("DATA_TYPE:", self.DATA_TYPE)
        print("WEIGHT_WIDTH:", self.WEIGHT_WIDTH)
        print("WEIGHT_ROW:", self.WEIGHT_ROW)
        print("WEIGHT_COL:", self.WEIGHT_COL)
        print("INPUT_WIDTH:", self.INPUT_WIDTH)
        print("RESULT_WIDTH:", self.RESULT_WIDTH)
        print("CIM_WIDTH:", self.CIM_WIDTH)
        print("CIM_DEPTH:", self.CIM_DEPTH)
        print("CIM_SIZE:", self.CIM_SIZE)
        print("IS_WIDTH:", self.IS_WIDTH)
        print("IS_DEPTH:", self.IS_DEPTH)
        print("OS_WIDTH:", self.OS_WIDTH)
        print("OS_DEPTH:", self.OS_DEPTH)
        print("IS_SIZE:", self.IS_SIZE)
        print("OS_SIZE:", self.OS_SIZE)
        print("FREQ:", self.FREQ)
        print("BUS_WIDTH:", self.BUS_WIDTH)
        print("PIPELINE_STAGES:", self.PIPELINE_STAGES)
        print()
